src/alchemlyb/convergence/convergence.py

```diff
@@ -389,3 +399,122 @@ def A_c(series_list, precision=0.01, tol=2):
             d_R_c = sorted_array[-i] - sorted_array[-i - 1]
             result += d_R_c * sum(R_c_list <= element) / n_R_c
     return result
+
+
+def block_average(df_list, estimator="MBAR", num=10, **kwargs):
+    """Free energy estimate for portions of the trajectory.
+
+    Generate the free energy estimate for a series of blocks in time,
+    with the specified number of equally spaced points.
+    For example, setting `num` to 10 would give the block averages
+    which is the free energy estimate from the first 10% alone, then the
+    next 10% ... of the data.
+
+    Parameters
+    ----------
+    df_list : list
+        List of DataFrame of either dHdl or u_nk, where each represents a
+        different value of lambda.
+    estimator : {'MBAR', 'BAR', 'TI'}
+        Name of the estimators.
+        See the important note below on the use of "MBAR".
+    num : int
+        The number of time points.
+    kwargs : dict
+        Keyword arguments to be passed to the estimator.
+
+    Returns
+    -------
+    :class:`pandas.DataFrame`
+        The DataFrame with estimate data. ::
+
+               FE             FE_Error
+            0  3.016442       0.052748
+            1  3.078106       0.037170
+            2  3.072561       0.030186
+            3  3.048325       0.026070
+            4  3.049769       0.023359
+            5  3.034078       0.021260
+            6  3.043274       0.019642
+            7  3.035460       0.018340
+            8  3.042032       0.017319
+            9  3.044149       0.016405
+
+
+    .. versionadded:: 2.4.0
+
+    """
+    logger.info("Start block averaging analysis.")
+    logger.info("Check data availability.")
+    if estimator not in (FEP_ESTIMATORS + TI_ESTIMATORS):
+        msg = f"Estimator {estimator} is not available in {FEP_ESTIMATORS + TI_ESTIMATORS}."
+        logger.error(msg)
+        raise ValueError(msg)
+    else:
+        # select estimator class by name
+        estimator_fit = estimators_dispatch[estimator](**kwargs).fit
+        logger.info(f"Use {estimator} estimator for convergence analysis.")
+
+    # Check that each df in the list has only one value of lambda
+    for i, df in enumerate(df_list):
+        lambda_values = list(set([x[1:] for x in df.index.to_numpy()]))
+        if len(lambda_values) > 1:
+            ind  = [j for j in range(len(lambda_values[0])) if len(list(set([x[j] for x in lambda_values]))) > 1][0]
+            raise ValueError(
+                "Provided DataFrame, df_list[{}] has more than one lambda value in df.index[{}]".format(i, ind)
+            )
+
+    if estimator in ["BAR"] and len(df_list) > 2:
+        raise ValueError(
+            "Restrict to two DataFrames, one with a fep-lambda value and one its forward adjacent state for a "
+            "meaningful result."
+        )
+        
+    # Choose length of comparison trajectory
+    lx_lambdas = [len(x) for x in df_list]
+    if len(set(lx_lambdas)) > 1:
```

Comment:
I'm not quite sure if this shall be the default behaviour. I can see the rational for this. 
However, the data needs to be decorrelated before passing into this function so each lambda might have different number of data points and this behaviour might not be expected by everyone.
I think it would be better to have this as an option where one is to use the minimum and the other is just 1/10 of all the data points.

Reply:
The way alchemlyb has constructed the `forward_backward_convergence` function, a list of dataframes is passed to the function:
```
df_list = [
    df_1,
    def_2,
    ...
]
```
One choose a value for `num`, let's say 10. In line's 112 through 114 a fraction of the data is taken from each `df` and concatenated together:
```
        sample = []
        for data in df_list:
            sample.append(data[: len(data) // num * I])
```
So if `len(df_1) == 100` and `len(df_2) == 80`, then each free energy estimation will have 20% fewer states for the state represented in df_1 versus df_2.

I guess the real question is, when calculating free energy with MBAR or BAR, is it ok that the different states (w_r and w_f) are  different sizes? From the code in pymbar the answer seems to be no. I'll remove this, and update the documentation to make it more clear.

Reply:
Thank you for the detailed response, and I appreciate the clarification.

Regarding your question:

> I guess the real question is, when calculating free energy with MBAR or BAR, is it ok that the different states (w_r and w_f) are different sizes?

I think there might be some confusion here. While it's true that having different sample sizes for different lambda states can make some people uncomfortable, particularly because it gives more weight to states with more samples in the MBAR estimator, whether this is good or bad isn't entirely clear and might depend on the specific context.

I'm glad you've decided to remove the code in question, as it allows flexibility for users. If someone prefers that different lambda states have varying sample sizes, they can still input a list of dataframes with different lengths, and the function will handle them accordingly. On the other hand, if a user believes that each lambda state should have the same number of samples, they can input dataframes of equal length, ensuring that the function will provide an equal number of samples for each calculation.

My concern with the original code was that when the final point, which is supposed to use 100% of the data, ends up using a reduced dataset (due to equalizing the sample sizes across lambda states), it might yield different results compared to simply feeding 100% of the data into the estimator. This could potentially lead to discrepancies that users might not expect.